lib/extensions.js

```diff
@@ -80,37 +80,43 @@ module.exports = function(realmConstructor, context) {
     const { DefaultNetworkTransport } = require('realm-network-transport');
     realmConstructor._networkTransport = new DefaultNetworkTransport();
 
-    Object.defineProperty(realmConstructor.Object.prototype, "toJSON", {
-        value: function () {
-            const result = {}
-            Object.keys(this).forEach(p => result[p] = this[p]);
-            Object.keys(Object.getPrototypeOf(this)).forEach(p => result[p] = this[p]);
-            return result;
+    Object.defineProperty(realmConstructor.Collection.prototype, "toJSON", {
+        value: function (_, cache = new Map()) {
+            return this.map((item, index) => item.toJSON ? item.toJSON(index.toString(), cache) : item);
         },
 
         writable: true,
         configurable: true,
         enumerable: false
     });
 
-    Object.defineProperty(realmConstructor.Object.prototype, "keys", {
-        value: function () {
-            return Object.keys(this).concat(Object.keys(Object.getPrototypeOf(this)));
-        },
+    Object.defineProperty(realmConstructor.Object.prototype, "toJSON", {
+        value: function (_, cache = new Map()) {
```

Comment:
Why do we have `cache` in the argument list. Is this intended to be passed in by the user and whats the idea. What can the user achieve with a custom map?

Reply:
`cache` has a default value (the `cache` is initialized if not supplied, and passed on internally), and thus is not required by the user (or `JSON.stringify` for that matter).
It's in the argument list for the `Realm.Object` & `Realm.Collection` implementations to share a cache for the current execution only.
`toJSON()` is a built-in functionality in JS. It's not intended to be called by the user, but can be (and should be able to be) called without arguments.

Reply:
sure, but why is it in the arguments list. Can it be declared inside the function body? Or I am missing something on this one.

Reply:
`Realm.Collection` & `Realm.Object` call `toJSON` on each other, and has to build up the cache together, for that execution only. This was the most elegant way I could think of.

Reply:
I guess I want to ask in that case, could we make this to default to null/undefined and instantiate it only where its used externally.

Reply:
ok ignore this. I am resolving this comment